zomes/chat/tests/batching.rs

```diff
@@ -1,98 +1,290 @@
+                tokio::runtime::Handle::current().block_on(shared_test_state.run(test_input));
+                Ok(())
+            })
+        }
+    })
+    .await
+    .unwrap();
+
+    Arc::try_unwrap(shared_test_state)
+        .unwrap_or_else(|_| panic!("shared test state should not have outstanding references because test has completed"))
+        .conductor
+        .shutdown()
+        .await;
+
+    test_result.unwrap();
+}
+
+fn expected_messages(test_input: TestInput) -> Vec<String> {
+    fn same_hour(a: &Timestamp, b: &Timestamp) -> bool {
+        let a = DateTime::try_from(a).unwrap();
+        let b = DateTime::try_from(b).unwrap();
+        a.signed_duration_since(b).num_hours().abs() == 0 && a.time().hour() == b.time().hour()
+    }
+
+    let TestInput {
+        message_history: mut messages,
+        earliest_seen,
+        target_message_count,
+    } = test_input;
+    messages.retain(|m| m.timestamp < earliest_seen);
+    messages.sort_by_key(|m| m.timestamp);
+    let (only_included_if_same_hour, included) =
+        messages.split_at(messages.len().saturating_sub(target_message_count));
+    let earliest_included_hour = if let Some(m) = included.first() {
+        &m.timestamp
+    } else {
+        return Vec::new();
+    };
+    if let Some(different_hour_idx) = only_included_if_same_hour
+        .iter()
+        .rposition(|m| !same_hour(&m.timestamp, earliest_included_hour))
+    {
+        messages.drain(0..=different_hour_idx);
+    }
+
+    messages.into_iter().map(|m| m.content).collect()
+}
+
+#[test]
+fn expected_messages_works() {
+    assert_eq!(
+        expected_messages(TestInput {
+            message_history: vec![
+                FakeMessage {
+                    content: "0".to_owned(),
+                    timestamp: Timestamp::from(chrono::Utc.ymd(2022, 1, 1).and_hms(0, 0, 0))
+                },
+                FakeMessage {
+                    content: "1".to_owned(),
+                    timestamp: Timestamp::from(chrono::Utc.ymd(2022, 1, 1).and_hms(0, 0, 0))
+                }
+            ],
+            earliest_seen: Timestamp::from(chrono::Utc.ymd(2022, 1, 1).and_hms(0, 0, 0)),
+            target_message_count: 1
+        }),
+        Vec::<String>::new(),
+    );
+
+    assert_eq!(
+        expected_messages(TestInput {
+            message_history: vec![
+                FakeMessage {
+                    content: "0".to_owned(),
+                    timestamp: Timestamp::from(chrono::Utc.ymd(2022, 1, 2).and_hms(2, 0, 0)),
+                },
+                FakeMessage {
+                    content: "1".to_owned(),
+                    timestamp: Timestamp::from(chrono::Utc.ymd(2022, 1, 1).and_hms(1, 0, 0)),
+                },
+            ],
+            earliest_seen: Timestamp::from(chrono::Utc.ymd(2022, 2, 1).and_hms(0, 0, 0)),
+            target_message_count: 1
+        }),
+        vec!["0".to_owned()]
+    );
+
+    // TODO: The real implementation has a bug here
```

Comment:
This is one lead: the DNA in this branch returns `["13"]` instead of `[]` given these inputs. I didn't investigate because in our real use-cases we always uses `target_message_count: 20`, but it's possible that investigating this would reveal the deeper bug with the chunking code.